packages/runtime/container-runtime/src/opLifecycle/outbox.ts

```diff
@@ -0,0 +1,191 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { IContainerContext } from "@fluidframework/container-definitions";
+import { GenericError } from "@fluidframework/container-utils";
+import { MessageType } from "@fluidframework/protocol-definitions";
+import { PendingStateManager } from "../pendingStateManager";
+import { BatchManager } from "./batchManager";
+import { IOutboxOptions, IBatchProcessors, BatchMessage, IBatch } from "./definitions";
+
+export class Outbox {
+    private readonly attachFlowBatch: BatchManager;
+    private readonly mainBatch: BatchManager;
+    private readonly defaultAttachFlowSoftLimitInBytes = 64 * 1024;
+
+    constructor(
+        private readonly shouldSend: () => boolean,
+        private readonly pendingStateManager: PendingStateManager,
+        private readonly context: IContainerContext,
+        private readonly options: IOutboxOptions,
+        private readonly batchProcessors: IBatchProcessors,
+    ) {
+        // We need to allow infinite size batches if we enable compression
+        const hardLimit = this.options.compressionOptions?.minimumBatchSizeInBytes !== Infinity
+            ? Infinity : this.options.maxBatchSizeInBytes;
+        const softLimit = this.options.compressionOptions?.minimumBatchSizeInBytes !== Infinity
+            ? Infinity : this.defaultAttachFlowSoftLimitInBytes;
+
+        this.attachFlowBatch = new BatchManager({
+            hardLimit,
+            softLimit,
+        });
+        this.mainBatch = new BatchManager({
+            hardLimit
+        });
+    }
+
+    public get isEmpty(): boolean {
+        return this.attachFlowBatch.length === 0 && this.mainBatch.length === 0;
+    }
+
+    public submit(message: BatchMessage) {
+        if (!this.mainBatch.push(message)) {
```

Comment:
Isn't size check should be after compression is done (at least in today's world, where we have 1Mb limit, but it applies to compressed content)? I think this breaks current behavior.
Maybe it's fine with assumption that we are going to not have these limits at all (at least for main batches).
But that makes it harder to have options for attachment ops. I.e. current implementation assumes that we are packing some number of attachment ops into smaller set of batches, and this is mostly controlled by size of those batches after compression. With this change it becomes before compression - it might be fine, but I think you will start running into issues where it may not fit into 1Mb after compression, and even attachment batches need to be chunked (probably need either way, as single attach op might be too large to require that anyway)

Reply:
All batches will be chuncked as needed, there's no difference between types of batches from this perspective.

> Isn't size check should be after compression is done (at least in today's world, where we have 1Mb limit, but it applies to compressed content)? I think this breaks current behavior.

The batch managers are configured with the proper limits based on whether or not compression is enabled, see https://github.com/microsoft/FluidFramework/pull/12938/files#diff-e3597b14f1a31ac595af232e2b96081c2a780dee3a142e9b517a24bdf5f7d704R27-R30 I don't see how this breaks existing behavior.

Reply:
this.options.maxBatchSizeInBytes should be after compression, I believe. and now it's before compression, right?
I think this limit (at least when it comes to regular batch) should be checked after we compressed - if compressed payload is over 1Mb, then it's game over, otherwise it should go through.

Attachment ops are harder, but somewhat similar logic should apply - ideally, we keep compression ops until we hit a (compressed limit), at which point we flush batch and start over.

Once chunking is in, the limit will move even further down the pipeline, as we should compress whole batch, and only then break it into chunks. We can do the same for attachment case (easier), but most performant solution would flush smaller batches as we accumulate enough ops (but do not need to wait for all attachment ops in this JS turn to be accumulated). We can punt on that optimization

Reply:
Here's my thinking right now:

- I will remove the attach batch manager from the compression/chunking logic. I think we can do smaller batches which we flush every once in a while, when the uncompressed payload size exceeds the configured limit. I think compression + chunking adds unnecessary overhead in this context. If it's safe to flush small batches of attach ops, we can leave it at that, the attach op batch size isn't the problem we're targeting right now. This will simplify the code here.
- the batch size for an IBatch object is irrespective of it being compressed or not. It's just the size of the batch as if we'd send it right now. What happens to it (like getting compressed or chunked) doesn't affect the semantics of the batch interface as it's oblivious to what has happened to it. The outbox will govern that aspect of the batches.
- for this particular line of code, there will be no change as we will always accept an op when compression is enabled, regardless of what the size of the current batch is (this is exactly what's happening right now in `main`). I will add another validation to check the size of the batch after compression and reject it (right now, if the compressed batch exceeds 1MB it will still be sent). This check will be replaced soon with chunking and that's when we will have no more limits.

Reply:
Let's chat about it

Reply:
I will ping you today

Reply:
We have discussed this offline. It would be best that when compression is enabled, in the case of the attach flow batch, don't wait for the next JS turn to flush. Instead, if the accumulated attach ops (pre compression) cross the compression size threshold, we should compress and flush at submit.